packages/flutter/lib/src/material/input_decorator.dart

```diff
@@ -1265,17 +1265,57 @@ class _RenderDecoration extends RenderBox {
     // See https://github.com/flutter/flutter/issues/13715
   }
 
+  double _lineWidth(double height, List<RenderBox?> boxes) {
+    double width = 0.0;
+    for (final RenderBox? box in boxes) {
+      if (box == null)
+        continue;
+      width = math.max(_minWidth(box, height), width);
+    }
+    return width;
+  }
+
   @override
   double computeMinIntrinsicHeight(double width) {
-    double subtextHeight = _lineHeight(width, <RenderBox?>[helperError, counter]);
+    final double iconHeight = _lineHeight(width, <RenderBox?>[icon]);
+    final double iconWidth = _lineWidth(iconHeight, <RenderBox?>[icon]);
+
+    width = math.max(width - iconWidth, 0.0);
+
+    final double prefixIconHeight = _lineHeight(width, <RenderBox?>[prefixIcon]);
+    final double prefixIconWidth = _lineWidth(prefixIconHeight, <RenderBox?>[prefixIcon]);
+
+    final double suffixIconHeight = _lineHeight(width, <RenderBox?>[suffixIcon]);
+    final double suffixIconWidth = _lineWidth(suffixIconHeight, <RenderBox?>[suffixIcon]);
+
+    width = math.max(width - contentPadding.horizontal, 0.0);
+
+    final double counterHeight = _lineHeight(width, <RenderBox?>[counter]);
+    final double counterWidth = _lineWidth(counterHeight, <RenderBox?>[counter]);
+
+    final double helperErrorAvailableWidth = math.max(width - counterWidth, 0.0);
+    final double helperErrorHeight = _lineHeight(helperErrorAvailableWidth, <RenderBox?>[helperError]);
+    double subtextHeight = math.max(counterHeight, helperErrorHeight);
     if (subtextHeight > 0.0)
       subtextHeight += subtextGap;
+
+    final double prefixHeight = _lineHeight(width, <RenderBox?>[prefix]);
```

Comment:
As I understand it, we want the layout logic here to roughly match what's in `performLayout`. For instance in `performLayout`:
https://github.com/flutter/flutter/blob/56c5f13035d3907a8c7586dee287265f35a236a2/packages/flutter/lib/src/material/input_decorator.dart#L974-L978

The `prefix` widget is laid out using the full width instead of the width of the text container. That could generate some layout difference I think. For example if the `prefix` widget is a `Text` widget that allows softwrap, then this `computeMinIntrinsicHeight` may return a larger height than needed.

Reply:
It seems a little bit strange that `prefix` `suffix` `prefixIcon` and `suffixIcon`  widgets are laid out using the full width, they are inside the container while icon is outside. For example if we look at field with prefix length greater than the field length:
```
TextField(
  maxLines: null,
  decoration: InputDecoration(
    labelText: 'Field with prefix',
    prefix: Text('PrefixPrefixPrefixPrefixPrefixPrefixPrefixPrefix'),
    icon: Icon(Icons.search),
    filled: true,
  ),
),
```
So the `prefix` goes outside the container:
<img width="404" alt="1" src="https://user-images.githubusercontent.com/74556185/131745980-42fb1ae0-a4e8-4e99-9e6a-80b1df9a27ce.png">
Probably we should change `_layout` in the part above to something like this:
```
boxToBaseline[icon] = _layoutLineBox(icon, boxConstraints);
final BoxConstraints containerConstraints = boxConstraints.copyWith(maxWidth: boxConstraints.maxWidth -  _boxSize(icon).width);
boxToBaseline[prefixIcon] = _layoutLineBox(prefixIcon, containerConstraints);
boxToBaseline[suffixIcon] = _layoutLineBox(suffixIcon, containerConstraints);
final BoxConstraints contentConstraints = containerConstraints.copyWith(maxWidth: containerConstraints.maxWidth - contentPadding.horizontal);
boxToBaseline[prefix] = _layoutLineBox(prefix, contentConstraints);
boxToBaseline[suffix] = _layoutLineBox(suffix, contentConstraints);
```
Then it's more correct:

Reply:
It seems a little bit strange that `prefix` `suffix` `prefixIcon` and `suffixIcon`  widgets are laid out using the full width, they are inside the container while icon is outside. For example if we look at field with prefix length greater than the field length:
```
TextField(
  maxLines: null,
  decoration: InputDecoration(
    labelText: 'Field with prefix',
    prefix: Text('PrefixPrefixPrefixPrefixPrefixPrefixPrefixPrefix'),
    icon: Icon(Icons.search),
    filled: true,
  ),
),
```
So the `prefix` goes outside the container (and outside the TextField actually):
<img width="404" alt="1" src="https://user-images.githubusercontent.com/74556185/131745980-42fb1ae0-a4e8-4e99-9e6a-80b1df9a27ce.png">
Probably we should change `_layout` in the part above to something like this:
```
boxToBaseline[icon] = _layoutLineBox(icon, boxConstraints);
final BoxConstraints containerConstraints = boxConstraints.copyWith(maxWidth: boxConstraints.maxWidth -  _boxSize(icon).width);
boxToBaseline[prefixIcon] = _layoutLineBox(prefixIcon, containerConstraints);
boxToBaseline[suffixIcon] = _layoutLineBox(suffixIcon, containerConstraints);
final BoxConstraints contentConstraints = containerConstraints.copyWith(maxWidth: containerConstraints.maxWidth - contentPadding.horizontal);
boxToBaseline[prefix] = _layoutLineBox(prefix, contentConstraints);
boxToBaseline[suffix] = _layoutLineBox(suffix, contentConstraints);
```
Then it looks like this:
<img width="400" alt="2" src="https://user-images.githubusercontent.com/74556185/131748322-a174068a-c477-4a61-85b5-88d1d7dd53b1.png">
And then `_layout` will be the same as `computeMinIntrinsicHeight`.
What do you think about it? @LongCatIsLooong @justinmc

Reply:
This seems the Android material_components library is allowing the `prefixText` `suffixText` to grow vertically without imposing any constraints  (I'm not familiar with Android layout primitives so I could be wrong): 
https://github.com/material-components/material-components-android/blob/34740b6fdf164665c438658935653d354a64d72b/lib/java/com/google/android/material/textfield/TextInputLayout.java#L779-L783

The solution you proposed makes sense. I couldn't find anything in the material spec regarding how prefix text and suffix text should be laid out.

Reply:
Can I change it now and this will be approved?